### PR TITLE
expression: implement vectorized evaluation for `builtinJSONMergeSig`

### PR DESCRIPTION
**Pull request by js00070**
### What problem does this PR solve? 
Implement vectorized evaluation for `builtinJSONMergeSig` mentioned in https://github.com/pingcap/tidb/issues/12104

### What is changed and how it works?
about 40% faster
```
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONMergeSig-VecBuiltinFunc-4                      482           2282113 ns/op          949025 B/op      14082 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONMergeSig-NonVecBuiltinFunc-4                   386           3197396 ns/op         1019904 B/op      14752 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_json_vec.go [hunk 1]
```diff
@@ -295,11 +296,67 @@ func (b *builtinJSONRemoveSig) vecEvalJSON(input *chunk.Chunk, result *chunk.Col
...
+		}(argBuffers[i])
+
+		if err := arg.VecEvalJSON(b.ctx, input, argBuffers[i]); err != nil {
+			return err
+		}
+	}
+
+	jsonValues := make([][]json.BinaryJSON, nr)
+
+	for i := 0; i < nr; i++ {
```
Comment (XuHuaiyu) [inline expression/builtin_json_vec.go]: Can we merge the 3 loops?

Comment (js00070) [reply expression/builtin_json_vec.go]: the second loop needs to run after the first loop, so does the third loop.

Comment (XuHuaiyu) [reply expression/builtin_json_vec.go]: Can we refine it like the following? If we want to reduce the cost of `runtime.growslice`, we can check `isNull` and then malloc jsonValues[i]. 
``` go
	result.ReserveJSON(nr)
OUTER:	
	for i := 0; i < nr; i++ {
		for j := 0; j < len(b.args); j++ {
			if argBuffers[j].IsNull(i) {
				result.AppendNull()
				goto OUTER
			}
			jsonValues[i] = append(jsonValues[i], argBuffers[i].GetJSON(j))
		}
		result.AppendJSON(json.MergeBinary(jsonValues[i]))
	}
```

Comment (js00070) [reply expression/builtin_json_vec.go]: I think this approch reduces the locality of memory accesses. In this approch, we have to go through all the argument columns row by row. I will do a benchstat to test it.

Comment (XuHuaiyu) [reply expression/builtin_json_vec.go]: ok

**Comment by js00070** [comment]
@XuHuaiyu 
I did several times of benchstat, the results are unstable because of my CPU condition(the frequency is changing on different temperature), but we can see my original code is faster than the approch like you mentioned.

```
name                                                                old time/op    new time/op    delta
VectorizedBuiltinJSONFunc/builtinJSONMergeSig-VecBuiltinFunc-12        822µs ± 1%    1064µs ± 0%  +29.47%  (p=0.000 n=19+17)
VectorizedBuiltinJSONFunc/builtinJSONMergeSig-NonVecBuiltinFunc-12    1.08ms ± 1%    1.30ms ± 1%  +20.19%  (p=0.000 n=19+20)
```
```
name                                                                old time/op    new time/op    delta
VectorizedBuiltinJSONFunc/builtinJSONMergeSig-VecBuiltinFunc-12        906µs ± 1%    1064µs ± 0%  +17.38%  (p=0.000 n=19+17)
VectorizedBuiltinJSONFunc/builtinJSONMergeSig-NonVecBuiltinFunc-12    1.18ms ± 1%    1.30ms ± 1%   +9.60%  (p=0.000 n=19+20)
```
```
name                                                                old time/op    new time/op    delta
VectorizedBuiltinJSONFunc/builtinJSONMergeSig-VecBuiltinFunc-12       1.02ms ± 1%    1.06ms ± 0%   +4.59%  (p=0.000 n=19+17)
VectorizedBuiltinJSONFunc/builtinJSONMergeSig-NonVecBuiltinFunc-12    1.32ms ± 1%    1.30ms ± 1%   -1.87%  (p=0.000 n=19+20)
```

the code which produces `new.txt` is as follow
```go
	result.ReserveJSON(nr)
	for i := 0; i < nr; i++ {
		nullFlag := false
		jsonVal := make([]json.BinaryJSON, 0, len(b.args))
		for j := 0; j < len(b.args); j++ {
			if argBuffers[j].IsNull(i) {
				result.AppendNull()
				nullFlag = true
				break
			}
			jsonVal = append(jsonVal, argBuffers[j].GetJSON(i))
		}
		if !nullFlag {
			result.AppendJSON(json.MergeBinary(jsonVal))
		}
	}
```

**Pull request merged**